### DISCUSSION TITLE
Make front-matter metadata of all pages available in jinja2

### DISCUSSION THREAD

**Discussion by Jc-L** [fralau/mkdocs-macros-plugin#110] [Q&A]
How to access to other pages' metadata to use in pure **jinja2** ? Use-case is to provide summary pages listing other pages matching some specific metadata (for example recently edited pages, pages marked for review, etc.

I attempted to use:

```
{% for page in navigation.pages %}
- {{ page }} / {{ page.title }} / {{ page.meta }}
{% endfor %}
```

However, in such metadata is only available for the current page, not for all pages.

Note: such info is available in vanilla mkdocs in the `pages` variable usable in templates. (see [pages.py](https://github.com/mkdocs/mkdocs/blob/c92eb40b5717a0af2cd44b738b608e02b6c3b9c2/mkdocs/structure/pages.py#L43)). The `pages` is a list of `Files` object that contain a `Pages` that contains the metadata.

**Comment by fralau** [comment]
Hi @Jc-L,

Sorry for my late answer. The specific jinja2 engine used by mkdocs_macros (on the markdown page) is **not** the same used by mkdocs (on HTML). The template engine used for mkdocs_macros is earlier on the page production chain.

Depending on what template engine you want to use, the answers are different (but they might also differ, according to what step of the macro conversion step you are one). You would have to tell me more precisely about your use case, and what effect you want to achieve?

**Comment by Jc-L** [comment]
Hi @fralau and thank you for the answer.

I am considering moving from a `Jekyll` static generator to `mkdocs`. I am familiar with `Jekyll` and make a significant use of metadata. For example, all my pages have a `status` metadata (draft, under_review, etc.), and I have a special page that lists all the pages where `status=draft` so that I can work on them.

This is just an example, and I considered the `macro` plugin to be a way to achieve this and other similar content management. I am however only digging the features of `mkdocs` and may have missed something here. Your mention of two engines running at different steps (markdown page and HTML page) makes me wonder if my approach is correct.

**Reply by fralau** [reply]
It metadata is all you need, this comes "out of the box" with mkdocs, using the [metadata extension of python-markdown.
](https://python-markdown.github.io/extensions/meta_data/)

You can put the metadata in the YAML header of each markdown page: this is beautifully documentated in the [Material theme](https://squidfunk.github.io/mkdocs-material/reference/) with examples -- but this can be applied to any theme.

To display the status on each page of the final website, you could typically 

1. Activate the metadata extension in your config (`mkdocs.yaml`) file.
2. Put the status info (e.g. `status: draft`) in the YAML header of your markdown file 
2. Create a HTML "block" containing this metadata, which would be inserted at the time of HTML generation.

You could also make use of the metadata in the YAML header using mkdocs-macros: but that would be if you want to use macros to make a _specific_ page to contain something related to metadata (or to something else). 

When python-markdown _extensions_ exist, it is not a bad idea to consider them as a first option (they basically extend the markdown syntax). There are two extension sets that I would recommend:

1. The [standard set for python-markdown](https://python-markdown.github.io/extensions/)
2. The magificent [Pymdown](https://facelessuser.github.io/pymdown-extensions/extensions/arithmatex/) set, that brings websites to the next level.

By contrast, the general idea of **mkdocs-macros** is to easily create effects that are not part of extensions, especially to cater for needs specific to a particular website, or because they require heaving lifting (e.g. extracting data out of a database, etc.). You will find a fuller [explanation on why and when mkdocs-macros here](https://mkdocs-macros-plugin.readthedocs.io/en/latest/why/).

**Reply by Jc-L** [reply]
Well, not sure things work like I expect. My example of a "summary" page containing metadata from other pages does not work, as only part of the other pages have their metadata populated when I use something like:

```
{% for page in navigation.pages %}
- {{ page }} / {{ page.title }} / {{ page.meta }}
{% endfor %}
```

**Reply by fralau** [reply]
Hmm... Since it seems an mkdocs question, my best recommendation is that you ask this question on [mkdocs's discussions](https://github.com/mkdocs/mkdocs/discussions)? It's very active and I am confident you will find some willing help.

**Reply by Jc-L** [reply]
Will do, Thank you for your time and support, and for the nice piece of software :)

**Reply by Jc-L** [reply]
After checking, I think the issue is not about mkdocs (no template usage, no HTML generation) but rather specific to the plugin. The issue being that `navigation.pages`  is only partially populated for the `meta` information of each page. I suppose this is because the parsing is done in the same loop as the reading of the metadata info. For my use-case to work, the initial loading of metadata should be done for all pages, and only then the parsing.

To reproduce, simply create two pages with:
```
{% for page in navigation.pages %}
- {{ page }} / {{ page.title }} / {{ page.meta }}
{% endfor %}
```

The rendered content will be different

I can check and propose a MR, if you think it is worth to.

**Reply by fralau** [reply]
I see. The philosophy is to keep the code of the plugin minimal. Perhaps you should write a [module](https://mkdocs-macros-plugin.readthedocs.io/en/latest/macros/) that does that, instead of modifying the code  of the plugin? 

Or, if you want to share it with other, a [pluglet](https://mkdocs-macros-plugin.readthedocs.io/en/latest/pluglets/)? Here is an [example of pluglet](https://github.com/fralau/mkdocs-macros-test).

**Reply by Jc-L** [reply]
Thank you for the tips. I understand the willingness to keep things simple (I am not sure it is a bug or not) and will see how to tackle the issue accordingly to the principles you mention.
